tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h

```diff
@@ -1009,199 +1009,22 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
-          // optimization, skip this integer divide if there's no dilation.
-          int64 lhs_spatial_index;
-          if (window_dim.base_dilation() > 1) {
-            lhs_spatial_index = undilated_index / window_dim.base_dilation();
-          } else {
-            lhs_spatial_index = undilated_index;
-          }
-
-          // Skip if input index is not in bounds.
-          if (!(lhs_spatial_index >= 0 &&
-                lhs_spatial_index < lhs_shape.dimensions(input_spatial_dim))) {
-            goto cnt;
-          }
-
-          lhs_linear_spatial_index +=
-              lhs_spatial_index * lhs_dim_multipliers[input_spatial_dim];
-          rhs_linear_spatial_index +=
-              (window_dim.window_reversal()
-                   ? ((window_dim.size() - 1) - rhs_spatial_index[ki])
-                   : rhs_spatial_index[ki]) *
-              rhs_dim_multipliers[dnums.kernel_spatial_dimensions(ki)];
-        }
-
-        for (int64 rhs_iz = 0; rhs_iz < input_feature_group_size; ++rhs_iz) {
-          const int64 iz =
-              feature_group_index * input_feature_group_size + rhs_iz;
-
-          int64 lhs_linear_index = lhs_linear_spatial_index;
-
-          lhs_linear_index += out_index[output_batch_dim] *
-                              lhs_dim_multipliers[input_batch_dim];
-
-          // We are scraping only the diagonal elements in the resultant
-          // convolution output when batch_group_count is greater than 1,
-          // where 1 is the default. No scraping is done in that case.
-          // This approach works out automatically for 'groups' in batches
-          // with group_size > 1, because we already descend down the batch
-          // dimension for the 'output_batch_dim' above.
-          lhs_linear_index +=
-              ((batch_group_index * batch_group_size) % input_batch_size) *
-              lhs_dim_multipliers[input_batch_dim];
-
-          lhs_linear_index += iz * lhs_dim_multipliers[input_z_dim];
-
-          int64 rhs_linear_index = rhs_linear_spatial_index;
-
-          rhs_linear_index += out_index[output_z_dim] *
-                              rhs_dim_multipliers[kernel_output_z_dim];
-          rhs_linear_index += rhs_iz * rhs_dim_multipliers[kernel_input_z_dim];
-
-          result_val +=
-              static_cast<ElementwiseT>(lhs_literal_data[lhs_linear_index]) *
-              static_cast<ElementwiseT>(rhs_literal_data[rhs_linear_index]);
-        }
-      cnt : {}
-      } while (IndexUtil::BumpIndices(window_shape,
-                                      absl::MakeSpan(rhs_spatial_index)));
-
-      return static_cast<ReturnT>(result_val);
-    };
-
-    Literal result(result_shape);
-    TF_RETURN_IF_ERROR(result.PopulateParallel<ReturnT>(func));
-
-    parent_->evaluated_[conv] = std::move(result);
-    return Status::OK();
-  }
+    if (result_element_type == input_element_type) {
+      return HandleConvolutionImpl<ReturnT, ElementwiseT>(conv);
+    }
+    return Unimplemented(
```

Comment:
Actually, @majnemer pointed out that other backends may not be able to implement convolutions with InputT=int8 while ResultT=float.

Overall:
* Before the int8 support, XLA only supports InputT = ResultT. All backends can implement this.
* With these patches, we enable XLA to support InputT = int8 && ResultT = float, which may not be implementable on some of our other backends.

In order to cope with this, I think we can introduce a slightly more verbose design:
* At XLA (e.g. opcode = kConvolution) level, add support for InputT=int8 and ResultT=in32.
* In the cudnn_*_conv_rewriter, pattern match `convert<dest=float>(conv<InputT=int8, ResultT=int32>(...))` and lower it to `custom-call<inputT=int8, ResultT=float`. For any ResultT=int32 that's not part of the pattern, return Unimplemented.

This means that the shape verification code in this patch is still legit, but the evaluator should be left untouched (assuming that it already supports InputT=int8 and ResultT=int32).

The rest of the patches should be fine (except for the rewriter that needs to change the pattern), because they all deal with xla/service/gpu.

Reply:
Originally, we decided to use CuDNN semantics for int8 convolution and introduced the int8-to-float convolution.  Int32 could be more applicable.  I agree to make the change to use int32 as a middle ground.  I had a ToDo comment about supporting int32 later.  Now I will flip or remove it.

Reply:
> Originally, we decided to use CuDNN semantics for int8 convolution and introduced the int8-to-float convolution.

I think the confusion comes from not distinguishing (a) backend-agnostic, XLA convolutions (before pattern matching) and (b) XLA GPU backend representation (after pattern matching).

We still follow our decisions for (b). However, for (a), we still need to consider the implement-ability of other backends. Hence I suggested to use int32 as output type for (a).

Reply:
Distinguishing (a) and (b) makes discussion easier.  I have the following suggestions for the design:
1. This patch has enabled int8-to-int32 convolution for (a).   I just need to take out the CuDNN specific comments.
2. I suggest not having int8-to-int8 convolution for (a) at this time.  It is allowed, but I don't think that's the purpose. It easily runs into overflow.  The overflow follows C language integer overflow semantics, which is undefined.  Since overflow needs many semantic parameters to define, maybe just defer it until it is requested.  Although int8-to-int32 also may see overflow, it is rare in practical use cases.  It is OK to me not to define the overflow semantics for int8-to-int32 convolution.
3. I suggest also allowing int8-to-float convolution for (a) by using int8-to-int32 convolution under the hood, because they match as long as int8-to-int32 convolution doesn't overflow.  Then, it is as same as convert<dest=float>(conv<InputT=int8, ResultT=int32>(...)).
4. For (b), cudnn_conv_rewriter allows int8-to-float, but not int8-to-int32.

I suggest 3. and 4. to keep int8-to-float for one major reason that we don't want people to use int8-to-int32 convolution and then supply an int32 bias by mistake, such as add<Input1T=int32, Input2T=int32>(conv<InputT=int8, ResultT=int32>(...)).  In this case, we can't fuse bias with convolution in one CuDNN call, because it doesn't match CuDNN semantics.  For backends that don't support int8-to-float convolution, they may disallow it as what CuDNN does for int8-to-int32.

Reply:
For (2), it's already supported for (a) even without your patches. If you feed an int8-to-int8 conv today, the verifier passes, and some backends may already implemented it.

> I suggest 3. and 4. to keep int8-to-float for one major reason that we don't want people to use int8-to-int32 convolution and then supply an int32 bias by mistake, such as add<Input1T=int32, Input2T=int32>(conv<InputT=int8, ResultT=int32>(...)). In this case, we can't fuse bias with convolution in one CuDNN call, because it doesn't match CuDNN semantics.

I agree that the pattern matching can be "slippery", aka users may write the wrong code and missed the fusion. In my general compiler(LLVM, XLA, etc) experience, I don't worry too much about it, as it's users responsibility to benchmark the application and tune the program to fit cuDNN when facing such a portability problem.

> For backends that don't support int8-to-float convolution, they may disallow it as what CuDNN does for int8-to-int32.

Correct, it's either adopt int8-to-float for (a) and let other backends disallow int8-to-float, or adopt int8-to-int32 for (a) and let Nvidia backend disallow int8-to-int32. To me I choose the latter because it costs less, as there are many other backends now and unbounded amount in the future.

To push this idea further, can we avoid changing (a) at all? Can we match the following pattern and use the existing int32-to-int32 for (a)?
```
cast<float>(conv<InputT=int32, ResultT=int32>(
    cast<int32>(x), cast<int32>(y)))
+ float_bias
```
?

Reply:
Using existing int32-to-int32 for (a) is a good idea.  Before making the switch, I would like check whether you have any int8 use case to share now, such as anything we must support or can't break.  The tests I add to test GPU backend have clear goal, but those for use case coverage are not quite.

Reply:
I don't have any int8 use case that can be shared publicly. I don't worry about the test coverage for now, since I'll run our internal tests when integrating the patches.

Reply:
@timshen91 , I got back working on the pattern matching implementation and just updated two commits.  The pattern matching solution only requires https://github.com/tensorflow/tensorflow/pull/30762 (no additional changes needed for pattern matching though), https://github.com/tensorflow/tensorflow/pull/30771 (all pattern matching implementation) , and https://github.com/tensorflow/tensorflow/pull/30783 (all tests for pattern matching and some misc changes), but not this one https://github.com/tensorflow/tensorflow/pull/30761 (as we avoid all semantics changes).  If you prefer, we may move our discussion to other PRs, as people won't find these discussions via code changes.

All pattern matching happens in https://github.com/tensorflow/tensorflow/pull/30783 in phases.  Here is a summary.  Detail patterns are listed in comments (mostly in the header file).
1. cudnn_conv_rewriter.cc matches 
    conv<InputT=int32, ResultT=int32>(convert<int32>(int8_x), convert<int32>(int8_y))
    and errors out if integer convolution is in other shape.
2. cudnn_fused_conv_rewriter.cc matches the rest:
2.1 RunFuseConvertToFloat absorbs the int32 to float cast following the convolution custom call, if any:
2.2 RunFuseBiasSideActivation does previous bias/side/activation fusion for all int8 and floating-point convolutions.
2.3 RunFuseClamp absorbs the clamp at the output.  It also absorbs the optional float to int8 cast at output and the int8 to float cast at side input, if both available.

I would like to get your early feedback.  I will add more tests to cover variations in addition to those https://github.com/tensorflow/tensorflow/pull/30783.